flotta-image-builder-common-functions.sh

```diff
@@ -0,0 +1,228 @@
+      make rpm
+  fi
+
+  # Build Prometheus node_exporter rpm
+  git clone https://github.com/project-flotta/prometheus-node_exporter-rpm.git /home/builder/prometheus-node_exporter-rpm
+  cd /home/builder/prometheus-node_exporter-rpm
+  make rpm
+
+  return
+}
+
+create_repo(){
+  PACKAGES_REPO=$1
+  dnf install -y createrepo httpd
+  systemctl enable --now httpd
+  firewall-cmd --zone=public --add-port=80/tcp --permanent
+  firewall-cmd --reload
+
+  mkdir /var/www/html/"$PACKAGES_REPO"
+  cp /root/rpmbuild/RPMS/${ARCH}/*.${ARCH}.rpm /var/www/html/"$PACKAGES_REPO"/
+  createrepo /var/www/html/"$PACKAGES_REPO"/
+
+  return
+
+}
+
+create_source(){
+  echo "creating source agent"
+  SOURCE_NAME=$1
+  PACKAGE_REPO_URL=$2
+
+  REPO_DIR=$(mktemp -d)
+  cat << EOF > $REPO_DIR/repo.toml
+id = "$SOURCE_NAME"
+name = "$SOURCE_NAME"
+description = "Flotta agent repository"
+type = "yum-baseurl"
+url = "$PACKAGE_REPO_URL"
+check_gpg = false
+check_ssl = false
+system = false
+EOF
+
+  composer-cli sources add $REPO_DIR/repo.toml
+  rm -rf $REPO_DIR
+
+}
+
+extract_image_to_web_server(){
+  BUILD_ID=$1
+  IMAGE_FOLDER=$2
+
+  echo "Saving image to web folder $IMAGE_FOLDER"
+  composer-cli compose image $BUILD_ID
+  sudo mkdir $IMAGE_FOLDER
+  sudo tar -xvf ${BUILD_ID//\"/}-commit.tar -C $IMAGE_FOLDER
+
+}
+
+create_kickstart_file(){
+  IMAGE_FOLDER=$1
+  ARCH=$(uname -i)
+  if [ "$ARCH" = "aarch64" ]; then
+      export OS_NAME="rhel-edge"
+      export REMOTE_OS_NAME="rhel-edge"
+      export REF="rhel/8/aarch64/edge"
+  else
+      export OS_NAME="rhel"
+      export REMOTE_OS_NAME="edge"
+      export REF="rhel/8/x86_64/edge"
+  fi
+
+  KICKSTART_TEMPLATE=edgedevice.ks.tmpl
+  KICKSTART_FILE=edgedevice.ks
+  echo "Creating kickstart file $KICKSTART_FILE"
+  envsubst < $KICKSTART_TEMPLATE > $KICKSTART_FILE
+  sudo cp $KICKSTART_FILE $IMAGE_FOLDER
+
+  echo "Image build completed successfully"
+}
+
+save_image_in_iso_format(){
+  IMAGE_SERVER_ADDRESS=$1
+  IMAGE_NAME=$2
+
+  cd /home/builder/r4e
+  ISO_BLUEPRINT=$(mktemp)
+  cat << EOF > $ISO_BLUEPRINT
+name = "edgedevice-iso"
+description = "Empty blueprint for ISO creation"
+version = "0.0.1"
+EOF
+
+  if [ "$ARCH" = "aarch64" ]; then
+      REF="rhel/8/aarch64/edge"
+  else
+      REF="rhel/8/x86_64/edge"
+  fi
+
+  # Save image in ISO format
+  dnf install -y jq isomd5sum genisoimage
+  composer-cli blueprints push $ISO_BLUEPRINT
+  BUILD_ID=$(composer-cli -j compose start-ostree --ref $REF --url http://$IMAGE_SERVER_ADDRESS/$IMAGE_NAME/repo/ edgedevice-iso edge-installer | jq '.build_id')
+  echo "waiting for build $BUILD_ID to be ready..."
+  while [ $(composer-cli -j compose status  | jq -r '.[] | select( .id == '$BUILD_ID' ).status') == "RUNNING" ] ; do sleep 5 ; done
+  if [ $(composer-cli -j compose status  | jq -r '.[] | select( .id == '$BUILD_ID' ).status') != "FINISHED" ] ; then
+      echo "image composition failed"
+      echo "check 'composer-cli compose status'"
+      exit 1
+  fi
```

Comment:
seems that this block is repeated several times and can be extracted into a common function.